betty/service.py

```diff
@@ -412,36 +418,79 @@ def _get(self, instance: _ServiceProviderT) -> _ServiceT:
         return new_service
 
 
-class _ServiceDecorator(Protocol):
-    @overload
-    def __call__(
-        self, factory: Callable[[_ServiceProviderT], _ServiceT]
-    ) -> _SynchronousServiceManager[_ServiceProviderT, _ServiceT]:
-        pass
+class _AsynchronousSharedServiceManager(
+    Generic[_ServiceProviderT, _ServiceT],
+    _SynchronousServiceManager[_ServiceProviderT, _ServiceT],
+):
+    async def initialize(self, instance: _ServiceProviderT) -> None:
+        setattr(
+            instance,
+            self._service_attr_name,
+            await self._get_factory(instance)(instance),
+        )
 
-    @overload
+
+class _AsynchronousServiceDecorator(Protocol):
     def __call__(
         self, factory: Callable[[_ServiceProviderT], Awaitable[_ServiceT]]
     ) -> _AsynchronousServiceManager[_ServiceProviderT, _ServiceT]:
         pass
 
 
+class _SynchronousServiceDecorator(Protocol):
+    def __call__(
+        self, factory: Callable[[_ServiceProviderT], _ServiceT]
+    ) -> _SynchronousServiceManager[_ServiceProviderT, _ServiceT]:
+        pass
+
+
+_ServiceDecorator: TypeAlias = (
+    _AsynchronousServiceDecorator | _SynchronousServiceDecorator
+)
+
+
+@overload
+def service(
+    factory: Callable[[_ServiceProviderT], Awaitable[_ServiceT]],
+    *,
+    shared: Literal[True],
+) -> _AsynchronousSharedServiceManager[_ServiceProviderT, _ServiceT]:
+    pass
+
+
 @overload
 def service(  # type: ignore[overload-overlap]
-    factory: Callable[[_ServiceProviderT], Awaitable[_ServiceT]], shared: bool = False
+    factory: Callable[[_ServiceProviderT], Awaitable[_ServiceT]],
+    *,
+    shared: Literal[False] = False,
 ) -> _AsynchronousServiceManager[_ServiceProviderT, _ServiceT]:
     pass
 
 
 @overload
 def service(
-    factory: Callable[[_ServiceProviderT], _ServiceT], shared: bool = False
+    factory: Callable[[_ServiceProviderT], _ServiceT],
+    *,
+    shared: bool = False,
 ) -> _SynchronousServiceManager[_ServiceProviderT, _ServiceT]:
     pass
 
 
 @overload
-def service(factory: None = None, shared: bool = False) -> _ServiceDecorator:
+def service(
+    factory: None = None,
```

Comment:
What if we remove support for no factories, and enforce `service` as a decorator only?